juju/juju
# [vsphere] Add ability to specify local templates for vsphere

This PR adds ability to specify local templates via simplestreams using the ```image-id``` format. I

## QA steps

Steps are as follows:

  * Identify the region you want to generate simplestreams for: juju regions yourcloud
  * Create a template for a series inside that particular datacenter. The template should have the same characteristics as the ones Juju creates if no simplestreams is supplied by the user. As a quick hack, simply deploy a controller, and clone the resulting template in a different location.
  * Generate simplestreams
  * Deploy a new controller using the generated simplestreams
  * Check that Juju skips downloading the template and moves straight to cloning the template.

### Generate tools metadata

```bash
mkdir -p $HOME/simplestreams/tools/proposed
cp $GOPATH/bin/jujud $HOME/simplestreams/tools/proposed
cd $HOME/simplestreams/tools/proposed
tar czf juju-`./jujud version`.tgz ./jujud
rm ./jujud
cd ~

juju-metadata generate-agents -d $HOME/simplestreams --stream proposed
```
### Generate image metadata

When generating simplestreams, use the relative path to the template in relation to the datacenter path (/datacenter-name/vm):

```bash
juju-metadata generate-image -d $HOME/simplestreams/ -i "gsamfira/juju-focal-template" -s focal -r datacenter6.5 -u 10.107.8.11
```

This will generate the following: https://paste.ubuntu.com/p/9zk5hX7wR6/

Given the above simplesrteams, the provider will look for the template in:

```bash
/datacenter6.5/vm/gsamfira/juju-focal-template
```

then deploy a new controller:

```bash
juju bootstrap --metadata-source $HOME/simplestreams --debug --verbose --show-log  vsphere vsphere-controller
```

## Bug reference
https://bugs.launchpad.net/juju/+bug/1867154
